### PR TITLE
Update bazel coverage test

### PR DESCRIPTION
**Pull request by oschaaf**
- Update the CI build image
- Update bazel coverage test
- Tweak in the python orchestration to allow
  for multiple test server instances
- Tweak in rate limiter testing of invalid input values,  which would
  otherwise misbehave in the new coverage test as-is.
- Switch build/test ci tasks to use gcc so we keep testing with it.

Greatly speeds up coverage tests as well as  produces more accurate
coverage percentages.

Needs #105 to land first

### CODE REVIEW

**Comment by oschaaf** [comment]
Just noticed in a local coverage run that this warning gets emitted:

```
warning: 15 functions have mismatched data
```

@htuch ^^ do you know if that is something we can safely ignore?

**Comment by htuch** [comment]
Not sure, @lizan might be more familiar with the current LLVM based sanitizer warnings.

**Comment by lizan** [comment]
should be OK, I didn't really dig it deeper than comparing coverage reports, though they seems to be from some dependencies.

**Comment by oschaaf** [comment]
Thanks @lizan, this is ready for review then
/assign htuch

File: ci/do_ci.sh [hunk 1]
```diff
@@ -5,11 +5,13 @@ set -e
 export BUILDIFIER_BIN="/usr/local/bin/buildifier"
 
 function do_build () {
+    setup_gcc_toolchain
     bazel build $BAZEL_BUILD_OPTIONS --verbose_failures=true //:nighthawk_client //:nighthawk_test_server \
-        //:nighthawk_service
+    //:nighthawk_service
```
Comment (htuch) [inline ci/do_ci.sh]: Nit: I think the indent was clearer before

File: ci/do_ci.sh [hunk 2]
```diff
@@ -5,11 +5,13 @@ set -e
...
 }
 
 function do_test() {
+    setup_gcc_toolchain
```
Comment (htuch) [inline ci/do_ci.sh]: FWIW, if we move to only clang we will be in a good place when it comes to both coverage and normal test/build; Envoy basically only uses gcc tests as a regression now, life is faster/better with Clang.

File: ci/do_ci.sh [hunk 3]
```diff
@@ -96,14 +107,11 @@ if [ -n "$CIRCLECI" ]; then
         mv "${HOME:-/root}/.gitconfig" "${HOME:-/root}/.gitconfig_save"
         echo 1
     fi
-
-    NUM_CPUS=8
-    if [ "$1" == "coverage" ]; then
-        NUM_CPUS=6
-    fi
+    # We constrain parallelism in CI to avoid running out of memory.
+    NUM_CPUS=6
```
Comment (htuch) [inline ci/do_ci.sh]: We shouldn't need to do this if we switch to Clang I think.

File: test/run_nighthawk_bazel_coverage.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  COVERAGE_THRESHOLD=98.3
+  COVERAGE_FAILED=$(echo "${COVERAGE_VALUE}<${COVERAGE_THRESHOLD}" | bc)
+  if test ${COVERAGE_FAILED} -eq 1; then
+      echo Code coverage ${COVERAGE_VALUE} is lower than limit of ${COVERAGE_THRESHOLD}
+      exit 1
+  else
+      echo Code coverage ${COVERAGE_VALUE} is good and higher than limit of ${COVERAGE_THRESHOLD}
+  fi
+fi
+echo "HTML coverage report is in ${COVERAGE_DIR}/index.html"
```
Comment (htuch) [inline test/run_nighthawk_bazel_coverage.sh]: I'm going to assume this is the same as the Envoy repo variant unless you have any diff you want to call out.

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 references:
   envoy-build-image: &envoy-build-image
-    envoyproxy/envoy-build:d0cefa7f071dbd4ef24399c2db8656c3a5d8c3ef
+    envoyproxy/envoy-build:8246167b9d238797cbc6c03dccc9e3921c37617d
```
Comment (htuch) [inline .circleci/config.yml]: @lizan do you think it would be easy/hard for us to move from CircleCI to Azure as well?

Comment (lizan) [reply .circleci/config.yml]: shouldn’t be hard especially if the project is not that large so you can do without RBE

File: ci/do_ci.sh (hunk 2 above)
Comment (oschaaf) [reply ci/do_ci.sh]: cool, done

File: ci/do_ci.sh (hunk 1 above)
Comment (oschaaf) [reply ci/do_ci.sh]: Fixed. It would be great to have check_format/fix for bash :)

File: ci/do_ci.sh (hunk 3 above)
Comment (oschaaf) [reply ci/do_ci.sh]: removed

File: test/run_nighthawk_bazel_coverage.sh (hunk 4 above)
Comment (oschaaf) [reply test/run_nighthawk_bazel_coverage.sh]: Yes, it's the same

**Comment by oschaaf** [comment]
comments addressed /assign htuch

**Comment by oschaaf** [comment]
@htuch I think we need to throttle parallelization even with clang, to avoid flakes in CI

**Comment by lizan** [comment]
@oschaaf seems we don't have do_circle_ci.sh in this repo which set NUM_CPUS=8, so yeah you need it somewhere.

**Comment by oschaaf** [comment]
Updated, we should now have `NUM_CPUS=8` set when we run in circleci

**Comment by oschaaf** [comment]
Yup CI is happy now that we capped parallelism at 8. Thanks again, @lizan !

**Pull request merged**